File: 0540-single-element-in-a-sorted-array/0540-single-element-in-a-sorted-array.java
class Solution {
    public int singleNonDuplicate(int[] nums) {

        int n = nums.length;
        int left = 0;
        int right = n-1;

        // checking for edge cases
        if(n==1){
            return nums[0];
        }else if(nums[0] != nums[1]){
            return nums[0];
        }else if(nums[right] != nums[right-1]){
            return nums[right];
        }

        while(left<=right){
            int mid = left + (right-left)/2;

            if(mid>0 && mid <n-1 && (nums[mid-1] != nums[mid] && nums[mid] != nums[mid+1])){
                return nums[mid];
            }

            // Intuition is if nums[mid] is at even index means it is the start sequence
            // if it is start sequence then mid+1 element should also be nums[mid]
            // if nums[mid] is in odd sequence then nums[mid-1] should be start sequence and both should be same
            // if this condition is satisfied then our non duplicate is not in left to mid range we have to search from mid + 1
            // similarly if this condition fails then it is in left to mid-1 range

            if(
                (mid %2 == 0 && nums[mid] == nums[mid+1]) 
                    ||
                (mid %2 == 1 && nums[mid] == nums[mid-1])
            ){ // move to right
                left = mid + 1;
            }else{
                right = mid - 1;
            }

        }

        return -1;
        
    }
}

File: 0540-single-element-in-a-sorted-array/0540-single-element-in-a-sorted-array.cpp
class Solution {
public:
    int singleNonDuplicate(vector<int>& nums) {

        int n = nums.size();
        int left = 0;
        int right = n-1;

        // checking for edge cases
        if (n==1){
            return nums[0];
        }else if (nums[0] != nums[1]){
            return nums[0];
        }else if (nums[right] != nums[right-1]){
            return nums[right];
        }

        while (left<=right){
            int mid = left + (right-left)/2;

            // check if mid is the non duplicate
            if(mid > 0 && mid < n-1 && nums[mid-1] != nums[mid] && nums[mid] != nums[mid+1]){
                return nums[mid];
            }

            // check if the mid is the start sequence or the end sequence
            // if it is start sequence then mid + 1 should also be same as nums[mid] (starts with even index)
            // if it is end sequene then mid -1 should be same as nums[mid] (ends with odd index)
            // if this is true then our non duplicate is not present in left to mid range we have to move to mid + 1
            // if false then we should see it in left to mid-1 range
            if (
                (mid %2 == 0 && nums[mid] == nums[mid+1]) || (mid %2 == 1 && nums[mid] == nums[mid-1])
            )
                {
                    // not present in left to mid range, move left to mid + 1
                
                    left = mid + 1;
                }else{
                    right = mid - 1;
                }
            
        }

        return -1;
        
    }
};